data_engineering.qmd

```diff
@@ -1,121 +1,526 @@
+
+2. **Setting Clear Objectives:**
+   The objectives for a KWS system might include:
+   - Achieving a specific accuracy rate (e.g., 98% accuracy in keyword detection).
+   - Ensuring low latency (e.g., keyword detection and response within 200 milliseconds).
+   - Minimizing power consumption to extend battery life on embedded devices.
+   - Ensuring the model's size is optimized for the available memory on the device.
+
+3. **Benchmarks for Success:**
+   Establish clear metrics to measure the success of the KWS system. This could include:
+   - True Positive Rate: The percentage of correctly identified keywords.
+   - False Positive Rate: The percentage of non-keywords incorrectly identified as keywords.
+   - Response Time: The time taken from keyword utterance to system response.
+   - Power Consumption: Average power used during keyword detection.
+
+4. **Stakeholder Engagement and Understanding:**
+   Engage with stakeholders, which might include device manufacturers, hardware and software developers, and end-users. Understand their needs, capabilities, and constraints. For instance:
+   - Device manufacturers might prioritize low power consumption.
+   - Software developers might emphasize ease of integration.
+   - End-users would prioritize accuracy and responsiveness.
+
+5. **Understanding the Constraints and Limitations of Embedded Systems:**
+   Embedded devices come with their own set of challenges:
+   - Memory Limitations: KWS models need to be lightweight to fit within the memory constraints of embedded devices. Typically, KWS models might need to be as small as 16KB to fit in the always-on island of the SoC. Moreover, this is just the model size. Additional application code for pre-processing may also need to fit within the memory constraints.
+   - Processing Power: The computational capabilities of embedded devices are limited (few hundred MHz of clock speed), so the KWS model must be optimized for efficiency.
+   - Power Consumption: Since many embedded devices are battery-powered, the KWS system must be power-efficient.
+   - Environmental Challenges: Devices might be deployed in various environments, from quiet bedrooms to noisy industrial settings. The KWS system must be robust enough to function effectively across these scenarios.
+
+6. **Data Collection and Analysis:**
+   For a KWS system, the quality and diversity of data are paramount. Considerations might include:
+   - Variety of Accents: Collect data from speakers with various accents to ensure wide-ranging recognition.
+   - Background Noises: Include data samples with different ambient noises to train the model for real-world scenarios.
+   - Keyword Variations: People might either pronounce keywords differently or have slight variations in the wake word itself. Ensure the dataset captures these nuances.
+
```

Comment:
Love the breakdown here! Another consideration that could be included is the specifications of the device used to record and collect this data. Differences in recording devices could also lead to unexpected biases in training data.

Reply:
Love the breakdown as well! I agree with Aditi fully that differences in recording devices is worth adding. This is especially noteworthy since its almost a guarantee that different groups will record with different devices.